ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/StreamedRow.java

```diff
@@ -181,31 +269,93 @@ private static void checkUnion(final Optional<?>... fs) {
   }
 
   @Immutable
+  public abstract static class BaseRow {
+
+    @Override
+    public String toString() {
+      try {
+        return OBJECT_MAPPER.writeValueAsString(this);
+      } catch (final JsonProcessingException e) {
+        return super.toString();
+      }
+    }
+  }
+
+  // Note: Type used `LogicalSchema` as a cheeky way of (de)serializing lists of columns.
+  @JsonInclude(Include.NON_EMPTY)
   @JsonIgnoreProperties(ignoreUnknown = true)
-  public static final class Header {
+  public static final class Header extends BaseRow {
 
     private final QueryId queryId;
-    private final LogicalSchema schema;
+    private final Optional<ImmutableList<SimpleColumn>> keySchema;
+    private final LogicalSchema columnsSchema;
 
-    @JsonCreator
     public static Header of(
-        @JsonProperty(value = "queryId", required = true) final QueryId queryId,
-        @JsonProperty(value = "schema", required = true) final LogicalSchema schema
+        final QueryId queryId,
+        final Optional<List<? extends SimpleColumn>> key,
+        final LogicalSchema columnsSchema
     ) {
-      return new Header(queryId, schema);
+      final Optional<List<? extends SimpleColumn>> keySchema = key
+          .map(k -> LogicalSchema.builder().valueColumns(k).build())
+          .map(LogicalSchema::columns);
+
+      return new Header(queryId, keySchema, columnsSchema);
     }
 
     public QueryId getQueryId() {
       return queryId;
     }
 
-    public LogicalSchema getSchema() {
-      return schema;
+    /**
+     * Used for push queries to return the schema of the key columns.
+     *
+     * <p>Note: The columns that make up the key may or may not be present in the projection, i.e.
+     * in the {@link #getColumnsSchema()}.
+     *
+     * @return the columns that make up the key.
+     */
+    @JsonIgnore
+    public Optional<List<SimpleColumn>> getKeySchema() {
+      return keySchema.map(v -> v);
     }
 
-    private Header(final QueryId queryId, final LogicalSchema schema) {
+    /**
+     * @return The schema of the columns being returned by the query.
+     */
+    @JsonProperty("schema")
+    public LogicalSchema getColumnsSchema() {
+      return columnsSchema;
+    }
+
+    @JsonProperty("key")
+    @SuppressWarnings("unused") // Invoked by reflection by Jackson.
+    private Optional<LogicalSchema> getSerializedKeySchema() {
+      // Type used Logical Schema to serialize key columns.
```

Comment:
Read this a few times but I'm still not understanding this comment. (The two lines below make sense but I can't parse `Type used Logical Schema to serialize key columns.` in a meaningful way.)

Reply:
I'll update.